### PR TITLE
hash: add more configurable hash algorithm for public key digest

### PR DESCRIPTION
**Pull request by dongx1x**
The current public key digest is using SHA1 and SHA256, and the digest will extend to TPM PCR 16. But from TPM2, SHA1 is not mandatory anymore, so this patch makes the hash algorithm configurable.

Reference:
TCG PC Client Platform TPM Profile Specification for TPM 2.0, Section 4.6 https://trustedcomputinggroup.org/wp-content/uploads/PC-Client-Specific-Platform-TPM-Profile-for-TPM-2p0-v1p05p_r14_pub.pdf

### CODE REVIEW

File: keylime/src/tpm.rs [hunk 1]
```diff
@@ -468,16 +465,11 @@ fn pubkey_to_tpm_digest<T: HasPublic>(
...
-    // SHA1
-    let mut hasher = openssl::sha::Sha1::new();
-    hasher.update(&keybytes);
-    let hashvec: Vec<u8> = hasher.finish().into();
-    keydigest.set(HashingAlgorithm::Sha1, Digest::try_from(hashvec)?);
+    let hash_algo = HashingAlgorithm::from(algo);
+    let mut hasher = Hasher::new(hash_alg_to_message_digest(hash_algo)?)?;
+    hasher.update(&keybytes)?;
+    let hashvec = hasher.finish()?;
+    keydigest.set(hash_algo, Digest::try_from(hashvec.as_ref())?);
```
Comment (ueno) [inline keylime/src/tpm.rs:472]: Previously, it extended PCR16 with both SHA-1 and SHA-256. With this change, the slot seems to be only extended with the given single algorithm. Wouldn't that be a problem?

Comment (dongx1x) [reply keylime/src/tpm.rs:472]: Hi @ueno, thanks for your comment! 
I think it's safe here, let me explain more, please correct me if I'm wrong.
The working flow of Keylime attestation should be: `extend PCR16` --> `build PCR list` --> `get a quote with PCR list` --> `check the quote` --> `check PCR 16`.

1. The PCR list is built with [the specific hash algorithm](https://github.com/keylime/rust-keylime/blob/master/keylime/src/tpm.rs#L311).

2.  When the verifier checks the PCR 16, the expected value is generated with [the specific hash algorithm](https://github.com/keylime/keylime/blob/master/keylime/tpm/tpm_main.py#L203).

So, although it extended PCR16 with both SHA-1 and SHA-256 but looks like only the specific one will be measured.

Comment (THS-on) [reply keylime/src/tpm.rs:472]: Yes, we check only the PCR 16 in the configured algorithm, so this change should not break anything.

**Review by THS-on — COMMENTED** [review]
@dongx1x can you fix the formatting (`cargo fmt`) and also add SM3_256 as an hash algorithm (that one is also supported by the server side: https://github.com/keylime/keylime/blob/master/keylime/common/algorithms.py)?

Otherwise the change LGTM.

File: keylime/src/tpm.rs [hunk 2]
```diff
@@ -452,10 +450,9 @@ fn parse_cred_and_secret(
 }
 
 // Takes a public PKey and returns a DigestValue of it.
-// Note: Currently, this creates a DigestValue including both SHA256 and
-// SHA1 because these banks are checked by Keylime on the Python side.
 fn pubkey_to_tpm_digest<T: HasPublic>(
     pubkey: &PKeyRef<T>,
+    algo: HashAlgorithm
```
Comment (aplanas) [inline keylime/src/tpm.rs:455]: nit: this variable is named `hash_alg` in other places

Comment (dongx1x) [reply keylime/src/tpm.rs]: OK, updated.

**Comment by dongx1x** [comment]
> @dongx1x can you fix the formatting (`cargo fmt`) and also add SM3_256 as an hash algorithm (that one is also supported by the server side: https://github.com/keylime/keylime/blob/master/keylime/common/algorithms.py)?
> 
> Otherwise the change LGTM.

Sure, updated.

**Review by aplanas — APPROVED** [review]
LGTM ... but I would wait the imput of @ansasaki or @ueno

**Review by ueno — APPROVED** [review]

**Review by ansasaki — APPROVED** [review]

**Pull request merged**